### PR TITLE
Add all instructions as a feature

### PR DESCRIPTION
**Pull request by boomanaiden154**
**Note:** This patch is currently based against the `added config options for vocab generation` patch, so a lot of the changes in this PR are related to that. The only relevant changes for adding the instruction features are the creation of the new model in `compiler_opt/rl/regalloc/config.py` at the top of the file and the adding of this model as a preprocessing layer for the `instructions_and_mapping` feature.

The current input tensors is a 34x300 matrix with the first row being the instruction opcodes and the next 33 rows being a binary matrix which maps each live range to its instructions. Eg:
```
op_code,op_code,op_code,op_code...
1            ,1            ,1            ,0           ...
0            ,0            ,0            ,1           ...
...
```

Currently the model performs the following steps on the input tensor:
1. Split the matrix into two parts, the first one being the instruction opcodes and the next one being the binary mapping matrix
2. Get embeddings for the instruction opcodes
3. Multiply the instruction embeddings by the binary matrix to get an output 33x16 matrix

Adding in a sparse optimizations on the matrix multiplication would probably be optimal, but tensorflow agents passes in the batch size as an additional dimension on the tensors passed to each preprocessing layer, and while the normal matrix multiplication algorithm supports having a batch size dimension, for some reason when enabling sparse optimizations, the function doesn't support having the batch size dimension. I'm planning on seeing how trivial this is to fix upstream and maybe writing a patch or just submitting an issue.

I'll try and post other potential models here that could potentially improve the performance of the model (since the matrix multiplication is extremely expensive) while still passing very similar/the same data to the model.

### CODE REVIEW

**Comment by boomanaiden154** [comment]
@yundiqian Here are the changes for the initial model plan. I'll try and get out an implementation of the idea that I mentioned on slack soon. Sorry for the delayed timeline. Validation that all the code worked as expected took longer than expected.

File: compiler_opt/rl/regalloc/config.py [hunk 1]
```diff
@@ -14,23 +14,68 @@
...
+from absl import logging
 
 def get_num_registers():
   return 33
 
+def get_opcode_count():
+  return 17716
+
+def get_num_instructions():
+  return 300
```
Comment (yundiqian) [inline compiler_opt/rl/regalloc/config.py:33]: where does this number come from? what if the 'affected #lines of code' is larger or smaller?

File: compiler_opt/rl/regalloc/config.py [hunk 2]
```diff
@@ -14,23 +14,68 @@
...
+    }
+
+  @classmethod
+  def from_config(cls, config):
+    return cls(**config)
+
+  def call(self, inputs):
+    # The first row is the instructions, and all rows afterwards compose the binary
+    # mapping matrix
+    instruction_opcodes = tf.slice(inputs, begin=[0, 0, 0], size=[-1, 1, self.instruction_count])
```
Comment (yundiqian) [inline compiler_opt/rl/regalloc/config.py]: so this is the op code for the 300 lines of code

File: compiler_opt/rl/regalloc/config.py [hunk 3]
```diff
@@ -14,23 +14,68 @@
...
+    instruction_opcodes = tf.reshape(instruction_opcodes, [-1, self.instruction_count])
+    binary_mapping_matrix = tf.slice(inputs, begin=[0, 1, 0], size=[-1, self.register_count, self.instruction_count])
```
Comment (yundiqian) [inline compiler_opt/rl/regalloc/config.py]: this is a very sparse matrix about which live range contains which line of code

File: compiler_opt/rl/regalloc/config.py [hunk 4]
```diff
@@ -14,23 +14,68 @@
...
+    binary_mapping_matrix_casted = tf.cast(binary_mapping_matrix, tf.float32)
+    instruction_embeddings = self.embedding_layer(instruction_opcodes)
+
+    matrix_product = tf.linalg.matmul(binary_mapping_matrix_casted, instruction_embeddings)
```
Comment (yundiqian) [inline compiler_opt/rl/regalloc/config.py:68]: this is the bag-of-word aggregation I mentioned

File: compiler_opt/rl/regalloc/config.py (hunk 1 above)
Comment (boomanaiden154) [reply compiler_opt/rl/regalloc/config.py:33]: It's just a number that I settled on after some experimentation. It seems to be able to encompass most of the instructions for every instance of the problem while not being too massively large. If the number of opcodes being sent is lower, the input gets zero-padded, and if the number of opcodes sent is larger, the data is just truncated.

File: compiler_opt/rl/regalloc/config.py (hunk 2 above)
Comment (boomanaiden154) [reply compiler_opt/rl/regalloc/config.py]: Yes. I'm passing everything in as a single `int64_t` matrix so that I can grab everything in one preprocessing layer and then splitting it up within the model (with this line grabbing the instruction opcodes).

File: compiler_opt/rl/regalloc/config.py (hunk 3 above)
Comment (boomanaiden154) [reply compiler_opt/rl/regalloc/config.py]: Yes. The whole input matrix (opcodes + mappings) is sparse, so it might even be best to just take a sparse tensor as input and convert the instruction opcodes vector to dense once it is extracted, but that point is moot due to the unavailability of a `sparse_dense_mat_mul` function that works with a batch size dimension.

File: compiler_opt/rl/regalloc/config.py (hunk 4 above)
Comment (boomanaiden154) [reply compiler_opt/rl/regalloc/config.py:68]: Yes. Just grabbing embeddings and then doing the matrix multiplication. There are a couple other options that I want to look at for the implementation of the bag of words aggregation that would in theory be more efficient than matrix multiplication. I'm going to try and get some code written up for that today.

**Comment by boomanaiden154** [comment]
```python
class process_instruction_features(tf.keras.Model):

  def __init__(self, opcode_count, register_count, instructions_per_live_range_count, embedding_dimensions=16):
    super().__init__()
    self.opcode_count = opcode_count
    self.register_count = register_count
    self.instructions_per_live_range_count = instructions_per_live_range_count
    self.embedding_dimensions = embedding_dimensions
    self.embedding_layer = tf.keras.layers.Embedding(self.opcode_count,
                                                     self.embedding_dimensions,
                                                     input_length=self.instructions_per_live_range_count * self.register_count)

  def get_config(self):
    return {
      "opcode_count": self.opcode_count,
      "register_count": self.register_count,
      "instructions_per_live_range_count": self.instructions_per_live_range_count,
      "embedding_dimensions": self.embedding_dimensions
    }

  @classmethod
  def from_config(cls, config):
    return cls(**config)

  def call(self, inputs):
    # inputs is a batch_size x register_count x instructions_per_live_range tensor containing opcodes.
    instruction_embeddings = self.embedding_layer(inputs)
    # we now have a batch_size, register_count, instructions_per_live_range, embedding_dimensions shaped tensor
    # sum along the instruction_count dimension
    output_tensor = tf.reduce_sum(instruction_embeddings, 2)
    return output_tensor
```
@yundiqian This was the other method I was describing on slack, where instead of taking a linear sequence of instructions and a mapping matrix, all of the instructions are just sent over as opcodes in a `num_registers * num_instructions_per_live_range` shaped tensor. Then an embedding layer is run on all of them and there is a reduction in the form of a sum/average. This should theoretically be faster than doing a full matrix multiplication, but I do want to make sure I'm not missing something.

**Comment by yundiqian** [comment]
This is a working solution for simple bag-of-word with only op code information about instructions. However, the issue with this representation is that:
1. information is duplicated X time (X == how many live ranges one instructions touches); it will be more severe if we want to include more per-instruction features other than op code in the future, e.g., frequency of this instruction.
2. the information of how different live ranges interference with each other gets lost, so we can hardly do anything more than the bag-of-word here

> ```python
> class process_instruction_features(tf.keras.Model):
> 
>   def __init__(self, opcode_count, register_count, instructions_per_live_range_count, embedding_dimensions=16):
>     super().__init__()
>     self.opcode_count = opcode_count
>     self.register_count = register_count
>     self.instructions_per_live_range_count = instructions_per_live_range_count
>     self.embedding_dimensions = embedding_dimensions
>     self.embedding_layer = tf.keras.layers.Embedding(self.opcode_count,
>                                                      self.embedding_dimensions,
>                                                      input_length=self.instructions_per_live_range_count * self.register_count)
> 
>   def get_config(self):
>     return {
>       "opcode_count": self.opcode_count,
>       "register_count": self.register_count,
>       "instructions_per_live_range_count": self.instructions_per_live_range_count,
>       "embedding_dimensions": self.embedding_dimensions
>     }
> 
>   @classmethod
>   def from_config(cls, config):
>     return cls(**config)
> 
>   def call(self, inputs):
>     # inputs is a batch_size x register_count x instructions_per_live_range tensor containing opcodes.
>     instruction_embeddings = self.embedding_layer(inputs)
>     # we now have a batch_size, register_count, instructions_per_live_range, embedding_dimensions shaped tensor
>     # sum along the instruction_count dimension
>     output_tensor = tf.reduce_sum(instruction_embeddings, 2)
>     return output_tensor
> ```
> 
> @yundiqian This was the other method I was describing on slack, where instead of taking a linear sequence of instructions and a mapping matrix, all of the instructions are just sent over as opcodes in a `num_registers * num_instructions_per_live_range` shaped tensor. Then an embedding layer is run on all of them and there is a reduction in the form of a sum/average. This should theoretically be faster than doing a full matrix multiplication, but I do want to make sure I'm not missing something.

**Comment by yundiqian** [comment]
for the high-dimension sparse dense matrix multiplication, i found this github issue about it: https://github.com/tensorflow/tensorflow/issues/9210

the most making-sense solution is mentions is https://github.com/tensorflow/tensorflow/issues/9210#issuecomment-850780779, but the author complains that it cannot calculate gradient. i tried a bit on it and it seems to be working well, including gradient computation:

```
import tensorflow as tf

st1 = tf.sparse.SparseTensor(indices=[[0, 0, 2], [1, 1, 1]], values=[10., 20.], dense_shape=[2, 3, 3])
dt1 = tf.Variable([[[1,2,3],[4,5,6],[7,8,9]],[[1,2,3],[4,5,6],[7,8,9]]], dtype=tf.float32)

def sparse_dense_matmult_batch(sp_a, b):
    return tf.map_fn(
        lambda x: tf.sparse.sparse_dense_matmul(x[0], x[1]),
        elems=(sp_a, b), fn_output_signature=tf.float32
    )

for _ in range(20):
  with tf.GradientTape() as tape:
    a = sparse_dense_matmult_batch(st1, dt1)
    b = tf.reduce_sum(a)
    loss = b * b
    tf.print(loss)

  grad = tape.gradient(loss, dt1)
  dt1.assign_sub(0.0001 * grad)
```

**Comment by boomanaiden154** [comment]
So I've done some more testing in regards to sparse matrices. I currently can't get the entire matrix to get passed to the `process_instruction_features` model as a sparse matrix (at least fully functionally, it does work, but the policy saver breaks) due to the fact that a `SparseTensorSpec` doesn't have a name attribute. I'll work on playing around with this a little bit more though because it is absolutely possible that I am just missing something. I also tried just converting the `binary_mapping_matrix` to a sparse matrix after it has been sliced and then passing it into the `sparse_dense_matmult_batch` function that you sent. However,  when doing it like this I'm seeing significantly reduced number of steps per second running behavioral cloning. I'm not totally sure why this is currently, but I'd hypothesize it's because `map_fn` isn't taking advantage of parallelism as efficiently as the non-sparse optimized `linalg.matmul` function, or that the conversion of the dense matrix into a sparse one is just taking up a significant amount of time. I'll continue to play around with it though and see if I can get something working.

**Comment by yundiqian** [comment]
ic, then i'm wondering how about just abandoning the sparse tensor concept. so the input becomes:
1. 300 * op_code instructions as before; later converted to 300 * 16 embedding table
2. 33 * 5 (or probably variable length, can start with a fixed number, e.g., 5) --- 5 being the #instructions this live range def/use, so for example it is [2, 5, 7, 9, 10] for live range 1, means live range 1 use/def instructions at line 2, 5, 7, 9, 10. Then you can use tf.gather to have 33 * 5 * 16 embedding table, then you do the tf.reduce_sum() to make it 33 * 16

in this way, it does the same thing as the sparse matrix multiplication, but does not contain sparse matrix

**Comment by boomanaiden154** [comment]
That should be pretty easily doable. I'll work soon on getting the necessary work done on the LLVM side to extract those features and then the ml-compiler-opt side should be pretty easy. I'll probably experiment with using all instructions that the live range encompasses (as I'm doing in this PR) as well as just use/def instructions as both should be fairly straightforward. It is also doesn't look too hard to add a name attribute to the `SparseTensorSpec` class, so I might experiment with that tonight and see if anything comes out of it. If it does, there might still be some issues to resolve (eg tensorflow version compatibility, seems to have already been resolved though), upstreaming, and then just waiting for a release cycle. Shouldn't hurt to at least try though.

**Comment by boomanaiden154** [comment]
Just keeping track of progress here. I hacked on `SparseTensor` a little bit last night and was able to pretty easily add a `name` attribute. However, just adding a name attribute doesn't push anything a whole lot farther as it doesn't get around the issue that all the logging on the LLVM side doesn't support sparse tensors. This is just for behavioral cloning, completely ignoring the whole problem of getting a saved model with a sparse tensor input working on the LLVM side (this would again require some rework on that side). Maybe something to look into supporting at some point, but definitely not a direction that is easy enough to pursue at this current time. Leaving the code changes in the `all-features-tooling-sparse-tensor` branch.

**Comment by yundiqian** [comment]
ic, i guess then we can just use the [2, 5, 7, 9, 10] index representations. the only issue with it is that it does not support variable lengths easily, but i guess it's fine for now.
